### PR TITLE
add context parallel group init to mp init

### PR DESCRIPTION
**Pull request by amylittleyang**
## What does this PR do?
follow up from [D55538929](https://www.internalfb.com/diff/D55538929).
update model_parallel/initialize.py to also initialize context parallel groups.
original change made in this PR: https://github.com/fairinternal/llm_inference/pull/333/files

## Before submitting

- [ ] Did you have fun?
  - Make sure you had fun coding 🙃
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/main/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/main/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: fairscale/nn/model_parallel/initialize.py [hunk 1]
```diff
@@ -88,32 +89,119 @@ def initialize_model_parallel(
...
+                    _PIPELINE_PARALLEL_GROUP = group
+                    _PIPELINE_PARALLEL_RANKS = ranks
+
+
+    # Build the context parallel groups.
+    global _CONTEXT_PARALLEL_GROUP
+    global _CONTEXT_PARALLEL_GROUP_RANKS
+    global _CP_PREV_RANK
+    global _CP_NEXT_RANK
+    global _CP_ZERO_RANK
```
Comment (jasonjk-park) [inline fairscale/nn/model_parallel/initialize.py:130]: wondering why we add these?
all these can be derived with CP GROUP RANKS (in fact, PP does that for similar functionalities)

we can always utilize @functools.lru_cache if performance is concern?

File: fairscale/nn/model_parallel/initialize.py [hunk 2]
```diff
@@ -88,32 +89,119 @@ def initialize_model_parallel(
...
+                    _CP_PREV_RANK = ranks[(index + 1) % len(ranks)]
+                    _CP_NEXT_RANK = ranks[index - 1]
+                    _CP_ZERO_RANK = ranks[0]
+
+
+def model_parallel_is_initialized() -> bool:
+    """Check if model and data parallel groups are initialized."""
+    if _MODEL_PARALLEL_GROUP is None or _DATA_PARALLEL_GROUP is None or _PIPELINE_PARALLEL_GROUP is None:
+        return False
+    return True
```
Comment (jasonjk-park) [inline fairscale/nn/model_parallel/initialize.py:153]: defined twice here and below?

also, should we add context parallel group to condition?

File: fairscale/nn/model_parallel/initialize.py [hunk 3]
```diff
@@ -88,32 +89,119 @@ def initialize_model_parallel(
...
+
+
+def get_context_parallel_group():
```
Comment (jasonjk-park) [inline fairscale/nn/model_parallel/initialize.py:156]: add return type annotation for all newly added functions?

File: fairscale/nn/model_parallel/initialize.py [hunk 4]
```diff
@@ -183,6 +271,5 @@ def destroy_model_parallel() -> None:
     _DATA_PARALLEL_GROUP = None
     global _PIPELINE_PARALLEL_GROUP
     _PIPELINE_PARALLEL_GROUP = None
-
-    global _PIPELINE_PARALLEL_RANKS
-    _PIPELINE_PARALLEL_RANKS = None
+    global _CONTEXT_PARALLEL_GROUP
+    _CONTEXT_PARALLEL_GROUP = None
```
Comment (jasonjk-park) [inline fairscale/nn/model_parallel/initialize.py:275]: should we also make all other CP based global variables None here?

File: fairscale/nn/model_parallel/initialize.py [hunk 5]
```diff
@@ -88,32 +89,119 @@ def initialize_model_parallel(
...
+    ), "context parallel group is not initialized"
+    return _CP_ZERO_RANK
+
+
+def get_context_parallel_world_size():
+    """Return world size for the context parallel group."""
+    if torch.distributed.is_available() and torch.distributed.is_initialized():
+        return torch.distributed.get_world_size(group=get_context_parallel_group())
+    else:
+        return 0
```
Comment (jasonjk-park) [inline fairscale/nn/model_parallel/initialize.py:196]: is 0 right default value? usually world size returns 1 by default?

File: fairscale/nn/model_parallel/initialize.py (hunk 1 above)
Comment (amylittleyang) [reply fairscale/nn/model_parallel/initialize.py]: Sounds good let me move this part into fbcode. It's probably for convenience in the original PR where they put all comms related changes in this file.

File: fairscale/nn/model_parallel/initialize.py (hunk 2 above)
Comment (amylittleyang) [reply fairscale/nn/model_parallel/initialize.py]: Good catch!

File: fairscale/nn/model_parallel/initialize.py (hunk 5 above)
Comment (amylittleyang) [reply fairscale/nn/model_parallel/initialize.py]: We should return torch.distributed.get_world_size(...) directly here. `get_context_parallel_group()` already checks the CP pg is initialized.

File: fairscale/nn/model_parallel/initialize.py [hunk 6]
```diff
@@ -88,41 +94,87 @@ def initialize_model_parallel(
...
+
+
+def get_context_parallel_world_size() -> int:
+    """Return world size for the context parallel group."""
+    return torch.distributed.get_world_size(group=get_context_parallel_group())
+
+
+def get_context_parallel_rank() -> int:
+    """Return my rank for the context parallel group."""
+    return torch.distributed.get_rank(group=get_context_parallel_group())
```
Comment (jasonjk-park) [inline fairscale/nn/model_parallel/initialize.py:175]: looks like line 161-165 is the same function - maybe remove this?

**Review by jasonjk-park — APPROVED** [review]

**Pull request merged**